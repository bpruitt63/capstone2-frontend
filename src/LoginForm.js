import React, {useState} from 'react';
import {Redirect, useHistory} from 'react-router-dom';
import {useHandleChange, useErrors} from './hooks';
import Errors from './Errors';
import BoateyApi from './BoateyApi';

function LoginForm({username, updateCurrentUser}) {

    const initialState = {username: '', password: ''}

    const [isLoading, setIsLoading] = useState(false);
    const [errors, setErrors] = useState({});
    const [data, handleChange, setData] = useHandleChange(initialState);
    const [apiErrors, getApiErrors, setApiErrors] = useErrors();
    const history = useHistory();

    /** Redirects to home if already logged in */
    if (username) {
        return <Redirect to='/' />
    };

    const handleSubmit = async (e) => {
        e.preventDefault();
        setErrors({});
        setApiErrors({});

        /** Validates login form and sets error if missing fields */
        if (!data.username || !data.password) {
            setErrors({error: "Username and Password are required"})
            setData(initialState);
            return false;
        } else {
            setIsLoading(true);

            /** Checks for valid name/password combination.
             * Returns API token.
             * Runs function to put username and token into state and local storage.
             */
            try {
                const {token, location, units} = await BoateyApi.login(data);
                updateCurrentUser(data.username, token, location, units);
                BoateyApi.token = token;
                history.push('/');
            } catch (e) {
                getApiErrors(e);
                setData(initialState);
                setIsLoading(false);
            };
        };
    };

    if (isLoading) {
        return <p>I'm loading!</p>
    };

    return (
        <>
            <Errors formErrors={errors}
                    apiErrors={apiErrors} />
            <form onSubmit={handleSubmit}>
                <label htmlFor='username'>Username</label>
                <input type='text'
                        name='username'
                        placeholder='Username'
                        value={data.username}
                        onChange={handleChange} />
                <label htmlFor='password'>Password</label>
                <input type='password'
                        name='password'
                        placeholder='Password'
                        value={data.password}
                        onChange={handleChange} />
                <button>Log In</button>
            </form>
        </>
    )
};

export default LoginForm;