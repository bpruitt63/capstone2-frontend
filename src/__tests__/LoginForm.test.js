import React from 'react';
import { render, screen } from '@testing-library/react';
import LoginForm from '../LoginForm';

test('renders', () => {
    render(<LoginForm isMobile={false}/>);
    expect(screen.getByText(/Log In/)).toBeInTheDocument;
    expect(screen.queryByText(/username/)).not.toBeInTheDocument;
});

test('renders mobile', () => {
    render(<LoginForm isMobile={true}/>);
    expect(screen.getByText(/Log In/)).toBeInTheDocument;
    expect(screen.queryByText(/username/)).toBeInTheDocument;
});

test('matches snapshot', () => {
    const {asFragment} = render(<LoginForm/>);
    expect(asFragment()).toMatchSnapshot();
});