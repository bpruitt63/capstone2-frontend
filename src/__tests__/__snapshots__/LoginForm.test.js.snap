// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="Errors"
    />
    <form
      class=""
    >
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <input
            class="form-control"
            id="username"
            name="username"
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
        <input
          class="mid form-control"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="button btn btn-secondary"
          >
            Log In
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
